server/src/main/java/eu/fasten/server/plugins/kafka/FastenKafkaPlugin.java

```diff
@@ -161,37 +168,84 @@ public void stop() {
     }
 
     /**
-     * Getter for the thread.
-     *
-     * @return thread
+     * Consumes a message from a Kafka topics and passes it to a plugin.
      */
-    public Thread thread() {
-        return thread;
+    public void handleConsuming() {
+        ConsumerRecords<String, String> records = connection.poll(Duration.ofSeconds(1));
+        Long consumeTimestamp = System.currentTimeMillis() / 1000L;
+
+        // Although we loop through all records, by default we only poll 1 record.
+        for (var r : records) {
+            processRecord(r, consumeTimestamp);
+        }
+
+        // Commit only after _all_ records are processed.
+        // For most plugins, this loop will only process 1 record (since max.poll.records is 1).
+        doCommitSync();
+
+        // If local storage is enabled, clear it after offsets are committed.
+        if (localStorage != null) {
+            localStorage.clear();
+        }
     }
 
     /**
-     * Consumes a message from a Kafka topics and passes it to a plugin.
+     * Consumer strategy (using local storage):
+     * <p>
+     * 1. Poll one record (by default).
+     * 2. If the record hash is in local storage:
+     * a. Produce to error topic (this record is probably processed before and caused a crash or timeout).
+     * b. Commit the offset, if producer confirmed sending the message.
+     * c. Delete record in local storage.
+     * d. Go back to 1.
+     * 3. If the record hash is _not_ in local storage:
+     * a. Process the record.
+     * b. Produce its results (either to the error topic, or output topic).
+     * c. Commit the offset if producer confirmed sending the message.
+     * d. Delete record in local storage.
+     * e. Go back to 1.
+     * <p>
+     * This strategy provides at-least-once semantics.
      */
-    private void handleConsuming() {
-        ConsumerRecords<String, String> records = connection.poll(Duration.ofSeconds(1));
-        for (var r : records) {
-            doCommitSync();
-            var consumeTimestamp = System.currentTimeMillis() / 1000L;
+    public void processRecord(ConsumerRecord<String, String> record, Long consumeTimestamp) {
+        if (localStorage != null) { // If local storage is enabled.
+            if (localStorage.exists(record.value())) { // This plugin already consumed this record before, we will not process it now.
+                logger.info("Already processed record with hash: " + localStorage.getSHA1(record.value()) + ", skipping it now.");
+                plugin.setPluginError(new ExistsInLocalStorageException("Record already exists in local storage. Most probably it has been processed before and the pod crashed."));
```

Comment:
Do you think that it might be a good idea to add the input record to the plugin's error in the case of `ExistsInLocalStorageException`? Consider a crash case where neither output nor error is produced, e.g. OOM.

Reply:
> Do you think that it might be a good idea to add the input record to the plugin's error in the case of `ExistsInLocalStorageException`? Consider a crash case where neither output nor error is produced, e.g. OOM.

@wzorgdrager any thoughts or ideas?

Reply:
Yes I agree, but I thought this was already the case when it was outputted to the error topic. I will take a look.

Reply:
The input is already added in this line (in case there is an exception) https://github.com/fasten-project/fasten/blob/fe64da43b641bca3c83848fffdf462c4ea822ab2/server/src/main/java/eu/fasten/server/plugins/kafka/FastenKafkaPlugin.java#L311 I will merge for now, to test on the cluster.